zorder/src/main/scala/org/locationtech/sfcurve/zorder/Z2.scala

```diff
@@ -97,54 +100,63 @@ object Z2 {
     *
     * @return (common prefix, number of bits in common)
     */
-  def longestCommonPrefix(lower: Long, upper: Long): (Long, Int) = {
-    var bitShift = MAX_BITS - MAX_DIM
+  def longestCommonPrefix(lower: Long, upper: Long): ZPrefix = {
+    var bitShift = TOTAL_BITS - MAX_DIM
     while ((lower >>> bitShift) == (upper >>> bitShift) && bitShift > -1) {
       bitShift -= MAX_DIM
     }
     bitShift += MAX_DIM // increment back to the last valid value
-    (lower & (Long.MaxValue << bitShift), MAX_BITS - bitShift)
+    ZPrefix(lower & (Long.MaxValue << bitShift), 64 - bitShift)
   }
 
-  /** Recurse down the quad-tree and report all z-ranges which are contained in the rectangle defined by the min and max points */
-  def zranges(min: Z2, max: Z2, globalMaxRecurse: Int = 32): Seq[IndexRange] = {
-    val (commonPrefix, commonBits) = longestCommonPrefix(min.z, max.z)
+  /**
+    * Recurse down the quad-tree and report all z-ranges which are contained
+    * in the cube defined by the min and max points
+    *
+    * @param min lower bound
+    * @param max upper bound
+    * @param precision bit precision of the z-values. can be used to stop searching after
+    *                  considering a certain number of bits.
+    * @return
+    */
+  def zranges(min: Z2, max: Z2, precision: Int = 64): Seq[IndexRange] = {
+    val ZPrefix(commonPrefix, commonBits) = longestCommonPrefix(min.z, max.z)
 
     // base our recursion on the depth of the tree that we get 'for free' from the common prefix
-    val maxRecurse = globalMaxRecurse
-
-    val mq = new MergeQueue
-    val sr = Z2Range(min, max)
-
-    var recCounter = 0
-    var reportCounter = 0
-
-    def _zranges(prefix: Long, offset: Int, quad: Long, level: Int): Unit = {
-      recCounter += 1
-
-      val min = prefix | (quad << offset) // QR + 000..
-      val max = min | (1L << offset) - 1  // QR + 111..
-
-      val nextLevel = level - 1
-      val qr = Z2Range(new Z2(min), new Z2(max))
-      if (sr containsInUserSpace qr){                         // whole range matches, happy day
-        mq += IndexRange(qr.min.z, qr.max.z, contained = true)
-        reportCounter +=1
-      } else if (sr overlapsInUserSpace qr) { // some portion of this range are excluded
-        if(offset > 0 && level > 0) {
-          _zranges(min, offset - MAX_DIM, 0, nextLevel)
-          _zranges(min, offset - MAX_DIM, 1, nextLevel)
-          _zranges(min, offset - MAX_DIM, 2, nextLevel)
-          _zranges(min, offset - MAX_DIM, 3, nextLevel)
-          //let our children punt on each subrange
+    val maxRecurse = if (commonBits < 30) 10 else if (commonBits < 40) 9 else 7
```

Comment:
i'll pull this out into a method - not sure we want constants, as the number of thresholds we set could vary. This was pulled from the z3 method (which was just experimentally estimated), but bumped up to maintain an approximately equal number of ranges based on 4 vs 8 sectors.